src/tiledb/cloud/vcf/ingestion.py

```diff
@@ -0,0 +1,977 @@
+
+        # vcf_attrs overrides extra_attrs
+        if vcf_attrs:
+            extra_attrs = None
+        elif isinstance(extra_attrs, str):
+            extra_attrs = [extra_attrs]
+
+        ds = tiledbvcf.Dataset(
+            dataset_uri, mode="w", cfg=tiledbvcf.ReadConfig(tiledb_config=config)
+        )
+        ds.create_dataset(
+            enable_allele_count=True,
+            enable_variant_stats=True,
+            extra_attrs=extra_attrs,
+            vcf_attrs=vcf_attrs,
+        )
+
+        # Create log array and add it to the dataset group
+        log_uri = f"{dataset_uri}/{LOG_ARRAY}"
+        create_log_array(log_uri)
+        group = tiledb.Group(dataset_uri, "w")
+        group.add(log_uri, name=LOG_ARRAY)
+        group.close()
+
+        write_log_event(log_uri, "create_dataset_udf", "create", data=dataset_uri)
+
+        create_manifest(dataset_uri)
+    else:
+        logger.info(f"Using existing dataset: '{dataset_uri}'.")
+
+    return dataset_uri
+
+
+def read_uris_udf(
+    dataset_uri: str,
+    list_uri: str,
+    *,
+    config: Optional[Mapping[str, Any]] = None,
+    max_files: Optional[int] = None,
+) -> Sequence[str]:
+    """
+    Read a list of URIs from a URI.
+
+    :param dataset_uri: dataset URI
+    :param list_uri: URI of the list of URIs
+    :param config: config dictionary, defaults to None
+    :param max_files: maximum number of URIs returned, defaults to None
+    :return: list of URIs
+    """
+
+    logger, profiler = setup(config, dataset_uri)
+
+    result = []
+    vfs = tiledb.VFS()
+    for line in vfs.open(list_uri):
+        result.append(line.decode().strip())
+        if max_files and len(result) == max_files:
+            break
+
+    logger.info(f"Found {len(result)} VCF files.")
+    profiler.finish()
+
+    return result
+
+
+def find_uris_udf(
+    dataset_uri: str,
+    search_uri: str,
+    *,
+    config: Optional[Mapping[str, Any]] = None,
+    pattern: Optional[str] = None,
+    ignore: Optional[str] = None,
+    max_files: Optional[int] = None,
+) -> Sequence[str]:
+    """
+    Find URIs matching a pattern in a directory or S3 bucket using the following command:
+
+        find <search_uri> -name <pattern> | grep -Ev <ignore> -m <max_files>
+
+    with an efficient implementation for S3.
+
+    :param dataset_uri: dataset URI
+    :param search_uri: URI to search for VCF files
+    :param config: config dictionary, defaults to None
+    :param pattern: pattern used in the search, defaults to None
+    :param ignore: exclude pattern applied to the search results, defaults to None
+    :param max_files: maximum number of URIs returned, defaults to None
+    :return: list of URIs
+    """
+
+    logger, profiler = setup(config, dataset_uri)
+
+    use_s3 = search_uri.startswith("s3://")
+    # Run command to find URIs matching the pattern
+    if use_s3:
+        cmd = f"exec aws s3 sync --dryrun --exclude '*' --include '{pattern}' {search_uri} ."
```

Comment:
Using `shell` here is vulnerable to accidental shell injection; instead the command should be built as a sequence:

```python
cmd = ("aws", "s3", "sync" "--dryrun", "--exclude", "*", "--include", pattern, search_uri, ".")
```

and then executed without `shell=True`.